keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -514,310 +481,126 @@ spec:
-that a claim gets allocated using that capacity while the kubelet is updating a
-`ResourceSlice`. The implementations of structured models must handle
-scenarios where more resources are allocated than available. The kubelet plugin
-of a DRA driver ***must*** double-check that the allocated resources are still
+**Note:** If a driver needs to remove a device or change its attributes,
+then there is a risk that a claim gets allocated based on the old
+`ResourceSlice`. The scheduler must handle
+scenarios where more devices are allocated than available. The kubelet plugin
+of a DRA driver ***must*** double-check that the allocated devices are still
 available when NodePrepareResource is called. If not, the pod cannot start until
-the resource comes back. Treating this as a fatal error during pod admission
+the device comes back. Treating this as a fatal error during pod admission
 would allow us to delete the pod and trying again with a new one.
 
 ### Using structured parameters
 
-The following is an example CRD which the developer of the
-`cards.dra.example.com` DRA driver might define as a valid claim parameters
-object for requesting access to its GPUs:
+A ResourceClaim is a request to allocate one or more devices. Each request in a
+claim may reference a pre-defined DeviceClass to narrow down which devices are
+desired or describe that directly in the request itself through a device
+selector. A device selector is a CEL expression that must evaluate to true if a
+device satisfies the request. A special `device` variable provides access to
+the attributes of a device.
 
-```yaml
-kind: CardParameters
-apiVersion: dra.example.com/v1alpha1
-metadata:
-  name: my-parameters
-  namespace: user-namespace
-  uid: foobar-uid
-...
-spec:
-  minimumRuntimeVersion: v12.0.0
-  minimumMemory: 32Gi
-  # "sharing" is a configuration parameter that does not
-  # get translated into the selector below.
-  sharing:
-    strategy: TimeSliced
-```
-
-Note that all fields in this CRD can be fully validated since it is owned by
-the DRA driver itself. This includes value ranges that are specific to the
-underlying hardware. There's no risk of using invalid attribute names because
-only the fields shown here are valid.
-
-With this CRD in place, a DRA driver controller is able to convert instances of
-it into a generic, in-tree `ResourceClaimParameters` object that the scheduler
-is able to understand.
+To correlate different devices, a claim may have "match attributes". Those are
+the names of attributes whose values must be the same for all devices that get
+allocated for the claim.
 
-For the example above, the converted object would look as follows:
+Configuration can be embedded in the claim for all devices at the claim level
+and separately for each request. These configuration parameters are ignored by
+the scheduler when selecting devices. They get passed down to the DRA drivers
+which provide the devices when preparing the devices on a node.
```

Comment:
In theory, we could pull configuration up to the claim-level only if we had a way to reference which requests a specific configuration applies to (similar to my proposal for matchAttributes).

Reply:
I'm undecided whether that makes the API simpler.

Reply:
I think it probably is the rare case, but it is a necessary case. I always go back to the MPS example, where I need to start an MPS server to govern a set of GPUs (that may be selected across different requests).

Reply:
Let's keep the common case simple, i.e. have configs associated with a request. The rare case is supported by also allowing configs associated with the claim.

This is the opposite of what we do with constraints, but there the common case (= apply to all devices) is the opposite, so doing it differently there is IMHO justified.

Reply:
OK, I still think it meakse sense to include a `RequestNames` here though, to mimic what we do with constraints and allow us to provide a config across a subset of requests.

Reply:
That's a good point.

I'm a bit worried that it'll become confusing where to put "per-device" configuration. It could go into the request or into the `claim.spec.config` with a single request name.

So perhaps only `claim.spec.config` after all?

@johnbelamaric ?

Reply:
I would be fine (possibly prefer) doing it only at the top.

Reply:
Let's  do just `claim.spec.config`, it mirrors `constraints`, and I expect the common case to be "no config".

Reply:
To be clear, I am suggesting `config.spec.config[*].requestNames` (or `deviceNames` as it were), listing the requests to which the config applies.

Reply:
Yes, I think we are all in agreement.

Reply:
Okay, changed.

Note that this also affects the `AllocationResult`.  I dropped "filtering": before, `RequestAllocationResult` contained only configuration parameters for the driver which provided the device in that request. Now `AllocationResult.Config` contains everything that was listed in claim or class and drivers need to pick out those parameters that are relevant for them and ignore the rest.

Reply:
The text above still implies it is embedded in the request object.

Reply:
Text fixed, but I am keeping this open for a bit longer until I have confirmation that the config restructuring is okay. Not a blocker, we can still revise if actually implementing it hits any snags.